src/awkward/operations/ak_from_buffers.py

```diff
@@ -129,7 +131,7 @@ def getkey(form, attribute):
 }
 
 
-def reconstitute(form, length, container, getkey, backend):
+def reconstitute(form, length, container, getkey, backend, simplify):
```

Comment:
I had to add a back door to `from_buffers` (notice that this only affects the implementation; it's inaccessible from high-level; we can make it public at any point in the future, if we have to, and then we'll think about the semantics).

The problem is that ArrayBuilder creates non-canonical Forms! (So the Form elements can't have the metadata constraint that Content nodes do; we have to support the existence of non-canonical Forms.) Specifically, if there's a union with missing values, ArrayBuilder makes `IndexedOptionArray(UnionArray(...))`. Actually, having ArrayBuilder _accumulate_ a canonical layout would make it a lot more complicated than it is; we should allow this to continue.

With this back door, `ak_from_buffers._impl` simplifies the layout while it builds it. This means that the resulting array will not have the same Form as is given to `ak_from_buffers._impl`, but there isn't a need for ArrayBuilder Forms to match the resulting array—it just has to make an array of the right type.

This back door should _not_ be what `ak.from_buffers` does because users would be very surprised if it normally produced arrays whose Form differs from the user-provided Form. The only reason we can get away with it with ArrayBuilder is because a human or third-party library is not involved in the construction of that Form.

We could make `ak.ArrayBuilder.form` canonicalize its own Form before showing it to the user by calling `.simplified` on its elements (recursively, from the leaves to the root). Then the secret will be safe...

Reply:
I'm thinking we might eventually expose it publicly, since it's possible to make non-canonical Forms. The thing to warn users about is that you'll get a different Form out than what you put in. Maybe `reform=False` (as the default)?